c-dependencies/js-compute-runtime/builtins/object-store.cpp

```diff
@@ -0,0 +1,464 @@
+
+    // Ensure that the NoGC is reset, so throwing an error in HANDLE_RESULT
+    // succeeds.
+    if (maybeNoGC.isSome()) {
+      maybeNoGC.reset();
+    }
+
+    if (!HANDLE_RESULT(cx, result)) {
+      return ReturnPromiseRejectedWithPendingError(cx, args);
+    }
+
+    int status =
+        fastly_object_store_insert(object_store_handle(self), key_chars, key_len, body_handle);
+    // Ensure that we throw an exception for all unexpected host errors.
+    if (!HANDLE_RESULT(cx, status)) {
+      return RejectPromiseWithPendingError(cx, result_promise);
+    }
+
+    // The insert was successful so we return a Promise which resolves to undefined
+    JS::RootedValue rval(cx);
+    rval.setUndefined();
+    JS::ResolvePromise(cx, result_promise, rval);
+    args.rval().setObject(*result_promise);
+
+    return true;
+  }
+
+  return false;
+}
+
+const JSFunctionSpec methods[] = {JS_FN("lookup", lookup, 1, JSPROP_ENUMERATE),
+                                  JS_FN("put", put, 1, JSPROP_ENUMERATE), JS_FS_END};
+
+const JSPropertySpec properties[] = {JS_PS_END};
+
+bool constructor(JSContext *cx, unsigned argc, JS::Value *vp);
+CLASS_BOILERPLATE(ObjectStore)
+
+bool constructor(JSContext *cx, unsigned argc, JS::Value *vp) {
+  REQUEST_HANDLER_ONLY("The ObjectStore builtin");
+  CTOR_HEADER("ObjectStore", 1);
+
+  JS::HandleValue name_arg = args.get(0);
+
+  size_t name_len;
+  // Convert into a String following https://tc39.es/ecma262/#sec-tostring
+  JS::UniqueChars name = encode(cx, name_arg, &name_len);
+  if (!name) {
+    return false;
+  }
+
+  // If the converted string has a length of 0 then we throw an Error
+  // because ObjectStore names have to be at-least 1 character.
+  if (name_len == 0) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr, JSMSG_OBJECT_STORE_NAME_EMPTY);
+    return false;
+  }
+
+  // If the converted string has a length of more than 255 then we throw an Error
+  // because ObjectStore names have to be less than 1025 characters.
+  if (name_len > 255) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr, JSMSG_OBJECT_STORE_NAME_TOO_LONG);
+    return false;
+  }
+
+  auto name_chars = name.get();
+
+  for (int i = 0; i < name_len; i++) {
+    char character = name_chars[i];
+    if (std::iscntrl(static_cast<unsigned char>(character)) != 0) {
+      JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                                JSMSG_OBJECT_STORE_NAME_NO_CONTROL_CHARACTERS);
+      return false;
+    }
+  }
+
+  ObjectStoreHandle object_store_handle = {INVALID_HANDLE};
+  int status = fastly_object_store_open(name_chars, name_len, &object_store_handle);
+  if (status == 2) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr, JSMSG_OBJECT_STORE_DOES_NOT_EXIST,
+                              name_chars);
+    return false;
+  }
+
+  if (!HANDLE_RESULT(cx, fastly_object_store_open(name_chars, name_len, &object_store_handle))) {
+    return false;
+  }
```

Comment:
Do we need to worry about closing the object store again, or is it just open for the duration of the request?

Reply:
I believe we do not need to worry about closing the object-store, which is the same approach taken for opening a dictionary. I checked and there are no close methods in the [ABI for object-store](https://github.com/fastly/ExecuteD/blob/main/wasm_abi/rust/fastly/src/object_store/handle.rs) or dictionary which makes me quite confident that we don't require closing them